content/blog/posts/2022-08-04-add-native-notifications.md

```diff
@@ -0,0 +1,172 @@
+---
+authorId: chuckcarpenter
+categories:
+  - notifications
+  - javascript
+date: ''
+slug: add-desktop-notifications
+title: Adding desktop notifications to your app/site
+---
+
+There are times when your application needs to inform users to act on something more
+timely than a normal notification (often known as toast notifications) may allow.
+If you need a user to do something and can't wait until they return to the application for them to notice or just want to engage them when something new is available, perhaps desktop notifications are a good idea for your app.
+
+This is a feature supported in all modern browers (sorry Internet Explorer, you're not invited). The user will have to opt in though and they will also have to allow notifications in their OS settings. Then you can notify them on the system level rather than just within your web browser page/tab.
+
+## What we'll work on today
+
+Let's create a basic notification that checks for and then gets the user's permission. Then we can show a notification about our cool new blog posts. We can then add some actions to allow the user to opt it future notifications as to when new posts are published. There's more to it, but we'll get to the nuances shortly.
+
+For this I'm going to create a singleton pattern file since I just need it to load and run when loaded. I find this to be a simple and readable pattern for our needs.
+
+```js
+class NotificationService {
+  constructor() {
+    // whatever you want once instance is created
+  }
+
+  async _isShowNotificationPermitted() {
+    // ...
+  }
+
+  async _showNotification() {
+    // ...
+  }
+}
+
+new NotificationService();
+```
+
+Using this javascript class I am adding a couple methods to handle checking the permission settings, requesting permissions, and then showing a notification if granted. Let's start with checking for and asking for permission to show notifications from our site.  
+
+```js
+async _isShowNotificationPermitted() {
+  if ('Notification' in window) {
+    let { permission } = Notification;
+    if (permission !== 'granted') {
+      permission =  await Notification.requestPermission();
+    }
+    if (permission !== 'denied') {
+      return true;
+    }
+  }
+
+  return false;
+}
+```
+
+I'll walk through the method a bit. First we're just basically checking for support in the browser in the off chance your user doesn't have a supported browser, otherwise why have those needless errors? The default value from `Notification.permission` is just that, `default`, which we treat essentially the same as `denied`. In that case we trigger `requestPermission()` in the API to request access from the user. If the user approves, we can move on to showing them notifications until this permission is revoked. Let's look then at a method for showing one.
```

Comment:
```suggestion
Let's break down this method a bit. First we're checking for support in the browser, in the off chance your user doesn't have a supported browser, otherwise why have those needless errors? The default value from `Notification.permission` is just that, `default`, which we treat essentially the same as `denied`. In that case we trigger `requestPermission()` in the API to request access from the user. If the user approves, we can move on to showing them notifications until this permission is revoked. Let's look then at a method for showing one.
```